### PR TITLE
Add reports/chart views

### PR DESCRIPTION
**Pull request by blms**
### Notes
- This PR only has one chart type enabled: bar charts that show breakdown of a single variable as a percentage of total enrollment. The others are [disabled in the dropdown](https://github.com/PREDICT-DPACC/dpdash/blob/2d9d22b5380abeb7979475550f844694e7c71bf1/views/components/Reports/ChartFormFields.jsx#L103-L107), and will [resolve to some placeholder text](https://github.com/PREDICT-DPACC/dpdash/blob/2d9d22b5380abeb7979475550f844694e7c71bf1/views/components/Reports/Chart.jsx#L9-L24) in a report.
- Enrollment targets are not yet supported.
- Date filtering is not yet supported on the frontend, but [it is on the backend](https://github.com/PREDICT-DPACC/dpdash/blob/2d9d22b5380abeb7979475550f844694e7c71bf1/server/routes/index.js#L1102-L1111) by supplying query parameters `start=YYYY-MM-DD` and `end=YYYY-MM-DD` to the request URL.

### Changes
- Add pages:
  - `/reports`
    - Reports list view (`/views/ReportsList*.react.js`, `/server/templates/ReportsList.template.js`)
  - `/reports/:id/view` 
    - Single report view (`/views/Report*.react.js`, `/server/templates/Report.template.js`)
  - `/reports/:id/edit`
    - Edit report view (`/views/EditReport*.react.js`, `/server/templates/EditReport.template.js`)
  - `/reports/new`
    - New report view (same as `EditReport`)
- Add API routes: 
  - [`/api/v1/studies/:study/enrollment`](https://github.com/PREDICT-DPACC/dpdash/wiki/API-documentation#apiv1studiesstudyenrollment)
  - [`/api/v1/reports`](https://github.com/PREDICT-DPACC/dpdash/wiki/API-documentation#apiv1reports)
  - [`/api/v1/reports/:id`](https://github.com/PREDICT-DPACC/dpdash/wiki/API-documentation#apiv1reportsid)
- Add helper component modules (`/views/components/Reports`) and utilities (`/views/fe-utils/reportsUtil.js`) for reports

### Testing

- [x] Rebuilt all code
- [x] Rebuilt Singularity image: `/data/predict/dpdash-ben-0622.sif`
- [x] Tested all affected functions in UI

### CODE REVIEW

**Comment by blms** [comment]
Added [user guide](https://github.com/PREDICT-DPACC/dpdash/wiki/User-guide:-Reports) to the wiki.

Tashrif's edit:
The wiki is easy to follow.

**Comment by tashrifbillah** [comment]
Is this branch already running on the dev server for me to test functionality?

**Comment by tashrifbillah** [comment]
Comments as I come along:

* (Minor) Remove `+` before ADD A CHART
* Although the legends are properly labeled for:
image

  Why are they not for the other variable?
  image
  
* (Feature request) DPdash can show saved graphs:
  
  image

  How hard would be to save the generated report so I don't have to generate it again next time I log in? Happy to discuss over a 
  call as well.

**Comment by blms** [comment]
> Why are they not for the other variable?

You have to fill out the "Title" field for the chart. (Between Chart Type and Assessment in the form)

> How hard would be to save the generated report so I don't have to generate it again next time I log in?

It should be doable, but let's make it a feature request in an issue instead of putting it into this PR.

**Comment by tashrifbillah** [comment]
> You have to fill out the "Title" field for the chart. (Between Chart Type and Assessment in the form)

I believe I shall have to do this:
image

**Comment by blms** [comment]
Oops, I misread! Missed where you said "legends". Yes, that's correct, you will have to fill it out like that.

**Comment by tashrifbillah** [comment]
> Date filtering is not yet supported on the frontend, but it is on the backend by supplying query parameters start=YYYY-MM-DD and end=YYYY-MM-DD to the request URL.

Can I get an example for the range `2021-05-20` to `2021-07-20`?

**Comment by blms** [comment]
> Can I get an example for the range `2021-05-20` to `2021-07-20`?

For total enrollment, it should be something like:
```
http://rc-predict-dev.partners.org/api/v1/studies/FAKE_1/enrollment?start=2021-05-20&end=2021-07-20
```
(And for enrollment by variable it's a `POST` request with that same URL and [`varName` and `assessment` in the body](https://github.com/PREDICT-DPACC/dpdash/wiki/API-documentation#post-2).)

However, this seems to be broken now, at least with my test just now using a `GET` request for total enrollment. So I'll troubleshoot that now.

File: stores/reducers.js [hunk 1]
```diff
@@ -1,22 +1,28 @@
 import { combineReducers } from 'redux';
 import { isBrowser } from '../server/utils/browserUtil';
 
-function subject(state = isBrowser() ? window.SUBJECT : null, action) {
```
Comment (tashrifbillah) [inline stores/reducers.js:4]: Why are `action` omitted?

**Comment by tashrifbillah** [comment]
Sorry, did not notice the https://github.com/PREDICT-DPACC/dpdash/wiki/API-documentation so far.


> So I'll troubleshoot that now.

Thank you. I am linking the [function](https://github.com/PREDICT-DPACC/dpdash/blob/1ad244b9a26c761bf51b3d860ad7c64159c1172c/server/utils/routerUtil.js#L129) here anyway.

File: stores/reducers.js (hunk 1 above)
Comment (blms) [reply stores/reducers.js:4]: They are unused, so I removed them to reduce confusion.

**Comment by blms** [comment]
Fixed the issue, it was a missing `await` keyword for the asynchronous function invocations.

Now this URL works:
```
http://rc-predict-dev.partners.org/api/v1/studies/FAKE_1/enrollment?start=2021-05-20&end=2021-07-20
```

**Comment by tashrifbillah** [comment]
Is this what I am supposed to see?
image

**Comment by blms** [comment]
Yes, as documented [here](https://github.com/PREDICT-DPACC/dpdash/wiki/API-documentation#get-2).

**Review by tashrifbillah — COMMENTED** [review]
* I verified they are there:
```vim
/reports
Reports list view (/views/ReportsList*.react.js, /server/templates/ReportsList.template.js)
/reports/:id/view
Single report view (/views/Report*.react.js, /server/templates/Report.template.js)
/reports/:id/edit
Edit report view (/views/EditReport*.react.js, /server/templates/EditReport.template.js)
```

* Some updates are also in `/views/components/`
* The new files are hooked up via `index.js` and `webpack.config.js`
* Most functionals are working at the UI

**Review by tashrifbillah — APPROVED** [review]
Ben can merge this PR after addressing some comments at https://github.com/PREDICT-DPACC/dpdash/issues/33.

**Pull request merged**